### PR TITLE
Property tests for path/git-ref validators and backend resolver (#70, #72)

### LINKED ISSUE #70: Property-based fuzz tests for safe_path / git_ref validators (cuprum/builders/args.py)
## Overview

Add Hypothesis property-based fuzz tests for `safe_path` and `git_ref` string validators in `cuprum/builders/args.py` (around line 52).

## Justification

String validators are classic fuzz targets; current tests cover only curated samples.

## Benefit

- Finds edge cases around empty strings, whitespace, NULs, path traversal, suffixes, and Unicode

## Cost Trade-Off

Low. Some oracle design needed.

## Verifiability Improvements

Factor validators into "classify reason" helpers so properties can assert on the rejection category rather than just pass/fail.

## Recommended Order

Phase 2 — validators and builders.

---
Raised from PR https://github.com/leynos/cuprum/pull/62 by @leynos.

### LINKED ISSUE #72: Property-based tests for _read_backend_env / get_stream_backend (cuprum/_backend.py)
## Overview

Add Hypothesis property-based tests for `_read_backend_env` and `get_stream_backend` in `cuprum/_backend.py` (around line 46).

## Justification

Env parsing and cached backend resolution combine normalisation, availability, and failure paths in ways difficult to cover with examples alone.

## Benefit

- Proves `AUTO` never leaks from the resolver
- Proves invalid values fail
- Proves forced Rust availability is respected

## Cost Trade-Off

Medium, due to cache and global environment resets.

## Verifiability Improvements

Split a pure `resolve_backend(requested, rust_available)` helper from the cached/env wrapper.

## Recommended Order

Phase 2 — validators and builders.

---
Raised from PR https://github.com/leynos/cuprum/pull/62 by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Adds Hypothesis property-based coverage for the argument validators (`safe_path`, `git_ref`) and the stream-backend resolver, backed by small refactors that expose pure, verifiable seams.

### `cuprum/builders/args.py` (#70)
- Introduce `PathRejection` and `GitRefRejection` enums whose member values are the exact error messages, and `classify_path_string` / `classify_git_ref` helpers that return the first failing category (or `None`).
- The validators now raise from the classifier, so the rejection *reason* is inspectable rather than pass/fail only. Existing behaviour and messages are unchanged.

### `cuprum/_backend.py` (#72)
- Split a pure `_resolve_backend(requested, rust_available)` decision from the cached, environment-reading, logging wrapper, plus `_probe_rust_availability` encapsulating each mode's probe-failure policy.
- All existing structured log events are preserved (resolution debug, forced-unavailable warning, auto-probe-failed debug).

### Tests
- `test_args_validators_property.py`: totality, validator/classifier consistency, and category coverage (NUL, `..` segments, whitespace, leading dash), plus round-trip/idempotence for valid inputs.
- `test_backend_resolver_property.py`: the resolver never leaks `AUTO`, honours `PYTHON`, respects forced-Rust availability, and env parsing rejects invalid values (only empty/`auto` resolve to `AUTO`).
- Regenerated the `test_maturin_build` wheel-manifest snapshot to include the two new modules.

## Verification

`make check-fmt lint typecheck test markdownlint` all green (779 Python items, 33 Rust nextest).

Closes #70
Closes #72

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by pandalump** [comment]
@coderabbitai review

File: cuprum/unittests/test_backend_resolver_property.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+
+@given(raw=_ENV_VALUE)
+def test_read_backend_env_rejects_invalid_values(raw: str) -> None:
+    """Env parsing yields a valid member or raises; only empty/auto give AUTO."""
+    normalised = raw.strip().lower()
+    valid = {member.value for member in StreamBackend}
+    prior = os.environ.get(_ENV_VAR)
+    os.environ[_ENV_VAR] = raw
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/unittests/test_backend_resolver_property.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: cuprum/builders/args.py [hunk 2]
```diff
@@ -119,4 +201,13 @@ def git_ref(value: str) -> GitRef:
...
 
 
-__all__ = ["GitRef", "SafePath", "git_ref", "safe_path"]
+__all__ = [
+    "GitRef",
+    "GitRefRejection",
+    "PathRejection",
+    "SafePath",
+    "classify_git_ref",
+    "classify_path_string",
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/builders/args.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: cuprum/unittests/test_args_validators_property.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    assert classify_path_string(raw, allow_relative=True) is (
+        PathRejection.PARENT_SEGMENT
+    )
+
+
+@given(segments=st.lists(_REF_SEGMENT, min_size=1, max_size=4))
+def test_absolute_paths_without_traversal_round_trip(segments: list[str]) -> None:
+    """Absolute, NUL-free, traversal-free paths validate and normalise stably."""
+    raw = "/" + "/".join(segments)
+    assert classify_path_string(raw, allow_relative=False) is None
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/unittests/test_args_validators_property.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

**Comment by pandalump** [comment]
@coderabbitai review

File: cuprum/_backend.py [hunk 4]
```diff
@@ -45,31 +45,58 @@ class StreamBackend(enum.StrEnum):
...
+    if not normalised:
         return StreamBackend.AUTO
     try:
-        return StreamBackend(raw)
+        return StreamBackend(normalised)
     except ValueError:
         valid = ", ".join(sorted(v.value for v in StreamBackend))
-        msg = f"invalid {_ENV_VAR} value {raw!r}; expected one of: {valid}"
+        msg = f"invalid {_ENV_VAR} value {normalised!r}; expected one of: {valid}"
         raise ValueError(msg) from None
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/_backend.py:81]: Keeping the full docstrings, deliberately, for consistency with this module's existing convention.

`cuprum/_backend.py`'s other private helpers already carry full numpy-style docstrings with `Returns`/`Raises`/`Notes` sections — e.g. `_read_backend_env` (Returns + Raises) and `_check_rust_available` (Returns + Notes), both predating this PR. Trimming only the two new private helpers to one-line summaries would make them inconsistent with their siblings in the same file, which reads as an oversight rather than an intentional style.

I checked for a governing rule and found none: neither `AGENTS.md` nor `.rules/*.md` mandates single-line docstrings for private functions. AGENTS.md instead states "Function documentation must include clear examples demonstrating usage and outcome," which is why `_resolve_backend` carries a doctest `Examples` block. The `Parameters`/`Returns`/`Raises` sections here document a non-obvious contract (`rust_available=None` semantics; the `ImportError` on forced-Rust-unavailable), so they earn their keep.

General rule I'd propose: match the established per-module docstring convention; a private helper's docstring depth should track its siblings and the subtlety of its contract, not a blanket line-count cap. Happy to revisit if a repo-wide "private = single-line" rule is adopted.

**Comment by pandalump** [comment]
@coderabbitai On the "User-Facing Documentation" pre-merge check:

The new `classify_path_string` / `classify_git_ref` classifiers and the `PathRejection` / `GitRefRejection` enums are a *developer* API (a testing/reasoning seam), not end-user behaviour — and they are documented, in `docs/developers-guide.md` under "Rejection classifiers for `safe_path` and `git_ref`", which records their intended audience and reuse policy. The user-facing behaviour of `safe_path` / `git_ref` is unchanged by this PR (same acceptances, same `ValueError` messages), so `docs/users-guide.md` has nothing new to state. developers-guide is the correct home for this API, per the repo's documentation-maintenance convention ("Internal interfaces: document ... in the relevant component/developers guide").

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai Resolved the "User-Facing Documentation" pre-merge check by taking the second option you offered — the classifiers are developer-facing only, so I have stopped exporting them.

`classify_path_string`, `classify_git_ref`, and the `PathRejection` / `GitRefRejection` enums are now removed from `cuprum/builders/args.py`'s `__all__`; its public surface is just `safe_path`, `git_ref`, and their `SafePath` / `GitRef` return types. The classifiers stay importable for in-tree callers and the property tests, and remain documented in `docs/developers-guide.md` (now noting they are intentionally out of `__all__`). Since they are no longer part of the public/user-facing API and no user-facing behaviour changed, `docs/users-guide.md` needs no new entry.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai Documented the backend resolution split in `docs/developers-guide.md` (new "Stream-backend resolution seam" subsection) describing `_parse_backend_value` / `_probe_rust_availability` / `_resolve_backend` and the single boundary handler.

These are private, internal helpers and the public `get_stream_backend` contract and precedence are unchanged, so developers-guide is the right home per the repo's documentation convention ("Internal interfaces: document ... in the ... developers guide"). `docs/cuprum-design.md` already documents the observable algorithm, which did not change, and a private-function decomposition of one function does not rise to an ADR — that would be disproportionate to this test-focused PR. Happy to reconsider if the backend selection design itself changes.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai The "Developer Documentation" pre-merge warning describes a gap that head `47088d1` already closes — the walkthrough appears to be evaluating the superseded commit `ff6ee30`. The single commit between them, `47088d1` ("Refine backend-seam docs: keyword-only sig and design-doc note"), adds exactly the two items the warning requests.

**1. Keyword-only marker for `_resolve_backend`** — `docs/developers-guide.md` (head `47088d1`) reads:

> - `_resolve_backend(requested, *, rust_available)` — the pure decision core
>   … (`rust_available` is keyword-only).

**2. Backend seam in the design doc** — `docs/cuprum-design.md` (head `47088d1`) now documents the seam:

> `_parse_backend_value` (pure env parsing), `_probe_rust_availability(requested)` (the impure probe), `_resolve_backend(requested, *, rust_available)` (the pure decision core, `rust_available` keyword-only).

The remaining references to `_check_rust_available()` are intentional: it is the coexisting cached availability probe (documented in `docs/adr-005-unified-rust-availability-probe.md`), a different function from the new resolution seam — not "older architecture."

`git grep "rust_available\` is keyword-only" 47088d1 -- docs/developers-guide.md` and `git grep _resolve_backend 47088d1 -- docs/cuprum-design.md` both confirm these are present on head, and absent on `ff6ee30`. Could you re-review against head `47088d1` and dismiss the stale warning?

File: docs/cuprum-design.md [hunk 5]
```diff
@@ -1869,6 +1869,15 @@ Cuprum selects the stream backend at runtime using the following precedence:
...
      other import failures after logging a warning.
+   - Internally, `get_stream_backend()` is factored into three seams for
+     testability without changing this precedence: `_parse_backend_value(raw)`
+     (pure env parsing), `_probe_rust_availability(requested)` (the impure
+     availability probe honouring each mode's failure policy), and
+     `_resolve_backend(requested, *, rust_available)` (the pure decision core,
+     which never returns `AUTO`). They are composed inside one boundary
+     `try`/`except`, so a forced-`rust` failure emits the
+     `cuprum.stream_backend_unavailable` warning whether the probe reports
+     unavailable or itself raises. See the developers' guide for details.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/cuprum-design.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai Addressed in `e6fd9b3`. Section 13.4 now presents one authoritative backend model:

- The prose leads with the resolution seam — `_parse_backend_value` → `_probe_rust_availability` → `_resolve_backend` — and describes `_check_rust_available()` as the cached probe that `_probe_rust_availability()` wraps, rather than as a competing delegation target.
- The mermaid flowchart routes the availability probe through a new `_probe_rust_availability()` node before `_check_rust_available()`, so the diagram matches the prose and the implementation (`cuprum/_backend.py`: `get_stream_backend` → `_probe_rust_availability` → `_check_rust_available`).

`make markdownlint` and `make nixie` pass. Could you re-review against the latest head and dismiss the Developer Documentation warning?

File: cuprum/unittests/test_backend_resolver_property.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            if result is StreamBackend.AUTO:
+                assert normalised in {"", "auto"}
+        else:
+            with pytest.raises(ValueError, match=_ENV_VAR):
+                _read_backend_env()
+    finally:
+        if prior is None:
+            os.environ.pop(_ENV_VAR, None)
+        else:
+            os.environ[_ENV_VAR] = prior
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_backend_resolver_property.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_args_validators_property.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+# absolute on POSIX but not on Windows (where it is drive-relative), so anchor
+# with a drive letter there to keep the round-trip property valid on both.
+_ABSOLUTE_ANCHOR = "C:/" if sys.platform == "win32" else "/"
+
+
+@given(raw=_FUZZ_TEXT, allow_relative=st.booleans())
+def test_classify_path_string_is_total(raw: str, *, allow_relative: bool) -> None:
+    """``classify_path_string`` returns a value or ``None`` but never raises."""
+    result = classify_path_string(raw, allow_relative=allow_relative)
+    assert result is None or isinstance(result, PathRejection)
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_args_validators_property.py:59]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_backend_resolver_property.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    valid = {member.value for member in StreamBackend}
+    if normalised == "":
+        # Empty/whitespace resolves to AUTO.
+        assert _parse_backend_value(raw) is StreamBackend.AUTO
+    elif normalised in valid:
+        # A recognized value maps to exactly that member, never a different one.
+        assert _parse_backend_value(raw).value == normalised
+    else:
+        with pytest.raises(ValueError, match=_ENV_VAR):
+            _parse_backend_value(raw)
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_backend_resolver_property.py:119]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/_backend.py [hunk 9]
```diff
@@ -242,19 +296,21 @@ def get_stream_backend() -> StreamBackend:
...
+            "Rust stream backend requested but unavailable",
+            extra={
+                "event": "cuprum.stream_backend_unavailable",
+                "requested_backend": requested.value,
+                "rust_available": rust_available,
+            },
+        )
+        raise
+    _log_stream_backend_resolution(requested, resolved, rust_available=rust_available)
+    return resolved
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/_backend.py:318]: Confirmed fixed on the current head. `get_stream_backend()`'s `try` now wraps **both** the probe and the resolver:

```python
rust_available: bool | None = None
try:
    rust_available = _probe_rust_availability(requested)
    resolved = _resolve_backend(requested, rust_available=rust_available)
except ImportError:
    _LOGGER.warning("Rust stream backend requested but unavailable", extra={...})
    raise
```

So both failure shapes — `_check_rust_available()` returning `False` (resolver raises) and the probe itself raising `ImportError` (genuine probe failure) — now emit the identical `cuprum.stream_backend_unavailable` warning with structured context (`rust_available` stays `None` in the latter case) before re-raising.

File: docs/developers-guide.md [hunk 10]
```diff
@@ -67,6 +67,51 @@ Both `TarCreateOptions` and `RsyncOptions` provide `allow_relative`. It
...
+and property tested in isolation:
+
+- `_parse_backend_value(raw)` — pure parsing of a `CUPRUM_STREAM_BACKEND` value
+  (whitespace/case normalisation, empty → `AUTO`, unknown → `ValueError`),
+  taking the raw string so tests inject values without mutating `os.environ`.
+- `_probe_rust_availability(requested)` — the impure availability probe,
+  encapsulating each mode's failure policy (`PYTHON` never probes; `AUTO`
+  tolerates a probe `ImportError`; forced `RUST` propagates it).
+- `_resolve_backend(requested, rust_available)` — the pure decision core that
+  never leaks `AUTO` and raises `ImportError` for forced-`RUST`-unavailable.
```
[the remark this answers is not in the document]
Comment (pandalump) [reply docs/developers-guide.md]: Confirmed fixed. `docs/developers-guide.md:106` documents the keyword-only form: `_resolve_backend(requested, *, rust_available)`, with a note that `rust_available` is keyword-only, so maintainers won't call it positionally.

**Comment by pandalump** [comment]
The post-merge Developer Documentation warning is valid: `docs/cuprum-design.md` §13.2 still implied `get_stream_backend()` delegates directly to `_check_rust_available()`. Since #224 is merged, I've addressed it in a follow-up: #248 qualifies the §13.2 sentence to route `get_stream_backend()` through the `_probe_rust_availability()` seam and cross-references §13.4's authoritative `_parse_backend_value` → `_probe_rust_availability` → `_resolve_backend` resolution. §13.3/§13.4 already described the seam model correctly, so this leaves one authoritative backend model. markdownlint + nixie pass.
